rust/src/rfb/rfb.rs

```diff
@@ -0,0 +1,799 @@
+            transactions: Vec::new(),
+            state: parser::RFBGlobalState::TCServerProtocolVersion
+        }
+    }
+
+    // Free a transaction by ID.
+    fn free_tx(&mut self, tx_id: u64) {
+        let len = self.transactions.len();
+        let mut found = false;
+        let mut index = 0;
+        for i in 0..len {
+            let tx = &self.transactions[i];
+            if tx.tx_id == tx_id + 1 {
+                found = true;
+                index = i;
+                break;
+            }
+        }
+        if found {
+            self.transactions.remove(index);
+        }
+    }
+
+    pub fn get_tx(&mut self, tx_id: u64) -> Option<&RFBTransaction> {
+        for tx in &mut self.transactions {
+            if tx.tx_id == tx_id + 1 {
+                return Some(tx);
+            }
+        }
+        return None;
+    }
+
+    fn new_tx(&mut self) -> RFBTransaction {
+        let mut tx = RFBTransaction::new();
+        self.tx_id += 1;
+        tx.tx_id = self.tx_id;
+        return tx;
+    }
+
+    fn get_current_tx(&mut self) -> Option<&mut RFBTransaction> {
+        for tx in &mut self.transactions {
+            if tx.tx_id == self.tx_id {
+                return Some(tx);
+            }
+        }
+        return None;
+    }
+
+    fn parse_request(&mut self, input: &[u8]) -> AppLayerResult {
+        // We're not interested in empty requests.
+        if input.len() == 0 {
+            return AppLayerResult::ok();
+        }
+
+        let mut current = input;
+        SCLogDebug!("request_state {}, input_len {}", self.state, input.len());
+        loop {
+            if current.len() == 0 {
+                return AppLayerResult::ok();
+            }
+            match self.state {
+                parser::RFBGlobalState::TSClientProtocolVersion => {
+                    match parser::parse_protocol_version(current) {
+                        Ok((rem, request)) => {
+                            current = rem;
+                            if request.major == "003" && request.minor == "003" {
+                                // in version 3.3 the server decided security type
+                                self.state = parser::RFBGlobalState::TCServerSecurityType;
+                            } else {
+                                self.state = parser::RFBGlobalState::TCSupportedSecurityTypes;
+                            }
+
+                            match self.get_current_tx() {
+                                Some(current_transaction) => {
+                                    current_transaction.ts_client_protocol_version = Some(request);
+                                }
+                                _ => {
+                                    return AppLayerResult::err();
+                                }
+                            }
+                        }
+                        Err(nom::Err::Incomplete(v)) => {
+                            if let nom::Needed::Size(n) = v {
```

Comment:
Are you sure the size `n` here indicates the size of the full record required? Or could it also be something like "I need 2 more bytes to get some integer"? In the latter case you should probably use `n + current.len()`. You must tell the API how much data you need on top of what was already consumed. So this needs to be true: `consumed + needed > input.len()'. Currently this is enforced by the API in a hard way (`BUG_ON` statements).

Reply:
Okay, got it. Then it should be `current.len() + n` indeed: `current.len()` for what is not consumed yet but was tried to parse, and `n` for what nom thinks it needs to complete on top of what is already has in `current`.
I assume I the `input` I get in the next call would then be guaranteed to be larger than `current.len() + n` from the previous call?